app/__init__.py

```diff
@@ -63,6 +64,10 @@ def refresh_session():
     application.add_template_filter(question_references)
     application.add_template_filter(parse_document_upload_time)
 
+    # the lucky thread that gets to do create_app also gets to pre-load its content_loader, in the process
+    # asserting that all required content files are present and in good order
+    get_content_loader()
+
     return application
```

Comment:
> the thread that manages to perform the init_app also gets to pre-load its content

I can't find any documentation on this. 

I was under the understanding that requests are initialised in a new thread in the same process with a new app context.

I may be missing when the threads are spawned though as the documentation on that is super sparse

Reply:
Yep the documentation properly sucks.

The threads actually get spawned by `uwsgi` (https://github.com/unbit/uwsgi/blob/b6308cae818dab78da5f51eae8c903b6e2122b7a/plugins/python/python_plugin.c#L1388) and are each just handed requests as they come. The "app init" code though, if you trace it back, gets called at import-time. Imports are locked and done serially in python so they work more or less "by magic" and you don't typically have to worry about the thread safety of things you do at import-time. When the first thread gets handed a request it will need to perform all the imports required to set up & access the "application" object.

When a second thread gets handed a request it will these imports will have already been performed and that thread will be able to access that module's attributes like a big dict. Normal module import behaviour really, just ensuring two threads can't be trying to execute import-time code at the same time.

Reply:
But really, the amount we're supposed to worry about this kind of stuff is "not much". The basic `uwsgi` flask examples really just seem to enable threads and fuhgettaboutit. We're only having to worry a bit because we've got mutable globals, which is generally a bad idea in a webapp altogether.